tools/src/tools.ml

```diff
@@ -310,6 +352,69 @@ let typeDetail typ ~env ~full =
          })
   | _ -> None
 
+(* split a list into two parts all the items except the last one and the last item *)
+let splitLast l =
+  let rec splitLast' acc = function
+    | [] -> failwith "splitLast: empty list"
+    | [x] -> (List.rev acc, x)
+    | x :: xs -> splitLast' (x :: acc) xs
+  in
+  splitLast' [] l
+
+let path_to_string path =
+  let buf = Buffer.create 64 in
+  let rec aux = function
+    | Path.Pident id -> Buffer.add_string buf (Ident.name id)
+    | Path.Pdot (p, s, _) ->
+      aux p;
+      Buffer.add_char buf '.';
+      Buffer.add_string buf s
+    | Path.Papply (p1, p2) ->
+      aux p1;
+      Buffer.add_char buf '(';
+      aux p2;
+      Buffer.add_char buf ')'
+  in
+  aux path;
+  Buffer.contents buf
+
+let valueDetail (typ : Types.type_expr) =
+  let rec collectSignatureTypes (typ_desc : Types.type_desc) =
+    match typ_desc with
+    | Tlink t | Tsubst t -> collectSignatureTypes t.desc
+    | Tconstr (Path.Pident {name = "function$"}, [t; _], _) ->
+      collectSignatureTypes t.desc
+    | Tconstr (path, ts, _) -> (
+      let p = path_to_string path in
+      match ts with
+      | [] -> [{path = p; genericParameters = []}]
+      | ts ->
+        let ts =
+          ts
+          |> List.concat_map (fun (t : Types.type_expr) ->
+                 collectSignatureTypes t.desc)
+        in
+        [{path = p; genericParameters = ts}])
+    | Tarrow (_, t1, t2, _) ->
+      collectSignatureTypes t1.desc @ collectSignatureTypes t2.desc
+    | Tvar None -> [{path = "_"; genericParameters = []}]
+    | Tpoly (t, ts) -> (
```

Comment:
@zth I'm not sure about `Tpoly` here.
Could you give an example of what this could be? Is that the thing with the `#type`?

Reply:
Thanks!